### PR TITLE
New infrastructure gateway based on `OrderingMpsc` and `UnorderedReceiver`

### PR DESCRIPTION
**Pull request by akoshelev**
This change glues together `OrderingMpsc[Sender|Receiver]` and `ChannelledTransport` inside the new gateway interface. Major changes are listed below.

* create new gateway interface and leverage `OrderingMpsc[Sender|Receiver]` and `UnorderedReceiver`
* clean up existing send and receive buffers, including fsv
* Remove the old `Transport` implementation and rename `ChannelledTransport` to `Transport`. 
* Infrastructure API has changed, therefore all protocols had to be upgraded to be compatible. 
* Fix imports of synchronization primitives (Arc, Mutex). Shuttle needs that to properly reorder tasks and make tests more reliable in discovering issues.

With proper buffer configuration, this scales up to 10k rows. It is hard to go above that (wait time is crazy - 40 mins for 20k rows, but IPA does not crash) because the impact of waking all tasks inside `OrderingMpscSender` starts to show up. Likely #522 is going to help a lot there.

This change also leaves HTTP/networking layer in a broken state for various reasons - PR is huge, in-memory implementation is more important to get out of the door. Work to enable e2e integration can be done in parallel with performance improvements.

Another major loss is the Infra's ability to print internal state when things go wrong. Will be added back once we settle with a particular send/receive implementation

### CODE REVIEW

**Review by martinthomson — COMMENTED** [review]
Lots of code to fix up after this, but maybe one more iteration on this is all we can really afford to spend.  Otherwise, the work to unblock this will get backed up.

File: src/helpers/gateway.rs
Comment (akoshelev) [reply src/helpers/gateway.rs:279]: that's actually `RWLock` quirk accidentally left over after migrating to DashMap. We don't need to use double-checked locking here, good catch!

File: src/helpers/gateway.rs
Comment (akoshelev) [reply src/helpers/gateway.rs:232]: shall we fix it when we merge both of these PRs together? I can totally do it then, currently it does seem consistent (both sender and receiver take capacity in number of messages)

Comment (martinthomson) [reply src/helpers/gateway.rs:232]: Fix it later, for sure.

**Review by martinthomson — APPROVED** [review]
Approved.  You'll need to rebase before landing though.

File: src/helpers/transport/mod.rs
Comment (andyleiserson) [reply src/helpers/transport/mod.rs:135]: A couple possibilities:
* If the choice is always known at compile time, use type aliases
* Make transport object-safe. There is a `dyn-clone` crate to support cloning (if we only ever have boxed instances and not references). The routing type parameters could probably be eliminated, either by taking an `Addr` instead, or serializing the route information, or some combination.

File: src/helpers/mod.rs
Comment (akoshelev) [reply src/helpers/mod.rs:80]: I like the idea of using `A`, `B` and `C`. I also agree with the concern of using indexing - `RoleAssignment` creation meant to hide it and helper identity config should probably do the same.

File: src/helpers/mod.rs
Comment (akoshelev) [reply src/helpers/mod.rs:341]: totally, I think step definition is correct and I don't want to make it `Copy` either. The annoyance is caused by the existence of `ChannelId` that owns `Step`. I tried to fix it but probably did not spend enough time on it - ideally I want to borrow the step from context and only clone it when I need to insert it inside the hashmap.

File: src/helpers/transport/mod.rs
Comment (akoshelev) [reply src/helpers/transport/mod.rs:135]: I think transport is actually object-safe, associated type is the problem here. I can make `Box<dyn T>>` but I have to constrain associated type as well, otherwise it won't compile. 

Looking at `TransportImpl` again, it basically implements type alias approach with quirks

File: src/helpers/transport/mod.rs
Comment (akoshelev) [reply src/helpers/transport/mod.rs:124]: I don't have the complete picture, but I was hoping we could just make HTTP transport implementation to hold a reference or own HTTP client. Each send should correspond to a new HTTP request with `data` attached as a body stream. I believe Hyper internally will spawn a task and poll it so we don't need to do it manually

File: src/test_fixture/world.rs
Comment (akoshelev) [reply src/test_fixture/world.rs:90]: Good question - I changed it because new Infra does not need to block anymore: all the magic happens inside the transport layer. I could change it back as I certainly cannot guarantee that we will never need to make it async again, but I also don't mind doing it - keeping things async where they don't need to be is a bit weird.

File: src/helpers/gateway/mod.rs
Comment (akoshelev) [reply src/helpers/gateway/mod.rs:74]: > I think it is because TransportImpl is an enum with no variants when the test-fixture feature is not set.

but we are not using it directly here, `send` is called on `RoleResolvingTransport`. Besides `TransportImpl` defines `send` so I am still puzzled why we have unused warning here

good point about the `Result` though - i missed it . Fixed

Comment (akoshelev) [reply src/helpers/gateway/mod.rs:74]: however `rustc` agrees with you

```bash
 cargo build --release --features test-fixture
```

gives me no warnings, but I don't understand why :)

**Pull request merged**
